### PR TITLE
Workaround for Prometheus sampleRing iterator bug 

### PR DESCRIPTION
**Pull request by rabenhorst**
Signed-off-by: Sebastian Rabenhorst <sebastian.rabenhorst@shopify.com>

There is a bug in Prometheus [sampleRing iterator](https://github.com/prometheus/prometheus/blob/211ae4f1f0a2cdaae09c4c52735f75345c1817c6/storage/buffer.go#L204-L378) for mixed native histogram types that repeats the last seen floatHistogram when followed by histograms and using [the `AtFloatHistogram` fn](https://github.com/prometheus/prometheus/blob/211ae4f1f0a2cdaae09c4c52735f75345c1817c6/storage/buffer.go#L280-L285). That results in zero rate when for example querying a native histogram over a downsampled (always type floatHistograms) and non downsampled block (type histogram.)

In this PR we add a workaround in the engine. It will be followed by PR for Prometheus.

### CODE REVIEW

File: engine/engine_test.go [hunk 1]
```diff
@@ -3340,6 +3340,49 @@ func TestNativeHistograms(t *testing.T) {
...
+		if i < 1 {
+			samples = append(samples, sample{t: ts, fh: th.ToFloat()})
+		} else {
+			samples = append(samples, sample{t: ts, h: th})
+		}
+	}
+
+	series := storage.NewListSeries(labels.Labels{{Name: "__name__", Value: "native_histogram_series"}}, samples)
+
+	// New Engine
```
Comment (fpetkovski) [inline engine/engine_test.go:3368]:
```suggestion
```

File: engine/engine_test.go [hunk 2]
```diff
@@ -3340,6 +3340,49 @@ func TestNativeHistograms(t *testing.T) {
...
+func testMixedNativeHistogramTypes(t *testing.T, opts promql.EngineOpts) {
+	test, err := promql.NewTest(t, "")
+	testutil.Ok(t, err)
+	defer test.Close()
+
+	var samples []tsdbutil.Sample
+
+	histograms := tsdbutil.GenerateTestHistograms(2)
+
+	for i, th := range histograms {
```
Comment (fpetkovski) [inline engine/engine_test.go:3357]: Let's use explicit indexes to access histograms. The `for` statement obfuscates what the code tries to do.

**Review by fpetkovski — COMMENTED** [review]
Looks good, just a few comments from my side.

File: execution/scan/matrix_selector.go [hunk 3]
```diff
@@ -256,13 +256,21 @@ loop:
 		switch buf.Next() {
 		case chunkenc.ValNone:
 			break loop
-		case chunkenc.ValHistogram, chunkenc.ValFloatHistogram:
-			t, h := buf.AtFloatHistogram()
+		case chunkenc.ValHistogram:
+			t, h := buf.AtHistogram()
```
Comment (fpetkovski) [inline execution/scan/matrix_selector.go:260]: We need the same fix here: https://github.com/thanos-community/promql-engine/blob/main/execution/scan/matrix_selector.go#L281

Comment (rabenhorst) [reply execution/scan/matrix_selector.go:260]: This should

> Looks good, just a few comments from my side.

I addressed the comments. Can you ptal.

**Review by fpetkovski — APPROVED** [review]
LGTM

**Comment by rabenhorst** [comment]
PR for error in distributed engine fuzz test (happens also in main): https://github.com/thanos-community/promql-engine/issues/229

**Review by fpetkovski — APPROVED** [review]
Thanks, this looks good now 👍

**Pull request merged**